[H4mes] iP

H4mes opened the pull request:
# Yarr: The Pirate Task Manager
```
Ahoy from
 _  _   __    ____  ____ 
( \/ ) /__\  (  _ \(  _ \
 \  / /(__)\  )   / )   /
 (__)(__)(__)(_)\_)(_)\_)
 ```
Have you ever felt:
- *Dangerously* forgetful?
- Bored of speaking plain English?
- Tired of minimalist app interfaces?

If you answered yes to any of the above, then Yarr is perfect for you! Yarr can be your swashbuckling assistant on the high seas of productivity.

## Getting Started with Yarr
To get acquainted with him:
1. download the most recent release from [here](https://github.com/H4mes/ip/releases);
2. run `java -jar yarr.jar` from your terminal
3. add your tasks
4. get chummy with your new favourite scallywag :pirate_flag:

All this for the low, low price of **FREE**!

## Features
Yarr continues to grow more capable each day and is able to:
- [X] Handle To-dos
- [X] Handle Deadlines and Events 
- [ ] Be operated with a GUI interface
- [ ] Accomodate flexible command formats

## Customisability
For enthusiastic programmers, Yarr is highly customisable with easily extended packages and classes. Here's a snippet of the `Task` class for example:
```java
public class Task {  
    protected String description;  
    protected boolean isDone;  
  
    public Task(String description) {  
        this.description = description;  
        this.isDone = false;  
    }  
  
    public String toString() {  
        return (isDone ? "[X] " + description : "[ ] " + description); //mark done task with X  
    }  
  
	public void markAsDone() {  
        isDone = true;  
    }
}
```
##
What are you waiting for? Download Yarr now and join our other happy users!
> The ~~One Piece~~ Productivity! The ~~One Piece~~ Productivity is real! – Edward Newgate

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

H4mes committed:
Level-0 first commit

H4mes committed:
Level-1 commit

H4mes committed:
Level-2 Commit

H4mes committed:
Removed TODO comments

H4mes committed:
Level-3 Commit

H4mes committed:
Level-4 Commit

H4mes committed:
A-TextUiTesting Commit

H4mes committed:
Level-5 Commit

H4mes committed:
Level-6 Commit

H4mes committed:
Format code to Java coding standard.

H4mes committed:
Add Level-7 requirements

Added loadData, appendTaskToFile, saveTaskListToFile, methods to allow
for saving and loading of the tasklist from the disk.

Added removeCorruptedData method to handle deletion of incorrect data at
the loadData stage. The method is designed to to only take an array of
indexes to allow for flexibility in what thrown exceptions are picked up
to be removed and to allow for deletion of larger numbers of corrupted
entries in a single pass.

Modified main method to use switch statement for todo, deadline and
event commands for better code readability.

H4mes committed:
Update Deadline, Event and Duke classes.

Updated Deadline and Event to store date as LocalDateTime.
Updated Deadline and Event getDate and getData methods to output the
correct date format.

Updated Duke with parseDate method to convert string in dd/MM/yyyy HHmm
format to LocalDateTime object.

H4mes committed:
Update Deadline, Event and Duke classes

Updated Deadline and Event to store date as LocalDateTime.
Updated Deadline and Event getDate and getData methods to output the
correct date format.

Updated Duke with parseDate method to convert string in dd/MM/yyyy HHmm
format to LocalDateTime object.

guohuang88 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,402 @@
+import java.io.*;
```
Perhaps it is better to import the specific class?

guohuang88 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,402 @@
+            ArrayList<Integer> corruptedLines = new ArrayList<>();
+            while (dataScanner.hasNextLine()) {
+                try {
+                    String line = dataScanner.nextLine();
+                    String[] data = line.split("\\|", 0);
+                    String taskType = data[0].trim();
+                    String taskComplete = data[1].trim();
+                    String taskDescription = data[2].trim();
```
I like how you name your variables as they are easy to understand! :smile:

guohuang88 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,24 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task{
+    protected LocalDateTime start;
+    protected LocalDateTime end;
+    protected DateTimeFormatter dataFormat = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
+    protected DateTimeFormatter stringFormat = DateTimeFormatter.ofPattern("MMM dd yyyy HHmm");
+    public Event(String description, LocalDateTime start, LocalDateTime end) {
+        super (description);
+        this.start = start;
+        this.end = end;
+    }
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + start.format(stringFormat) + " to: " + end.format(stringFormat) + ")";
+    }
+
+    @Override
+    public String getData() {
+        return "E | " + super.getData() + " | " + start.format(dataFormat) + " | " + end.format(dataFormat);
+    }
+}
```
Perhaps you can add JavaDoc here!

guohuang88 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
I like how you name your boolean here! 😄

guohuang88 commented on src/main/java/Duke.java:
Perhaps you can have more OOP here?

guohuang88 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,402 @@
+     * @param indexes List of line numbers (1-based indexing) to be deleted.
+     */
+    private static void removeCorruptedData(ArrayList<Integer> indexes) {
+        ArrayList<String> lines = new ArrayList<>();
```
I like how you use plural form for this variable! 💯

guohuang88 reviewed (commented):
Well done on your work! Overall, code is neat! 
Also, I have made some comments on the Java coding standard!

bachletuankhai commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+    protected DateTimeFormatter dataFormat = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
+    protected DateTimeFormatter stringFormat = DateTimeFormatter.ofPattern("MMM dd yyyy HHmm");
+    public Deadline(String description, LocalDateTime by) {
+        super (description);
```
Should the method name stay attached with `(`?

bachletuankhai commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
```
I think the [coding standard guide](https://se-education.org/guides/conventions/java/intermediate.html#comments) stated that you MUST write header comments for all classes and public methods. I notice that some of your classes and public methods do not have header comments.

bachletuankhai commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,402 @@
+                case "deadline": {
+                    String tempString = input.substring(9).trim();
+                    String[] tempArray = tempString.split("/by", 0);
+                    if (tempArray.length == 1) {
+                        throw new IllegalArgumentException("Blunder! " +
+                                "Declare yer deadline as such: 'deadline * /by *', ye scurvy dog!");
+                    }
+                    String description = tempArray[0].trim();
+                    LocalDateTime by = parseDate(tempArray[1].trim());
+                    printDivider();
+                    tasks.add(new Deadline(description, by));
+                    taskCount += 1;
+                    System.out.println("I've appended this to yer list:\n" + tasks.get(taskCount - 1).toString());
+                    printDivider();
+                    break;
+                }
+                case "event": {
+                    String tempString = input.substring(6).trim();
+                    String[] tempArray = tempString.split("/from", 0);
+                    if (tempArray.length == 1) {
+                        throw new IllegalArgumentException("Blunder! " +
+                                "Declare yer event as such: 'deadline * /from * /to *', ye scurvy dog!");
+                    }
+                    String description = tempArray[0].trim();
+                    tempString = tempArray[1].trim();
+                    tempArray = tempString.split("/to", 0);
+                    if (tempArray.length == 1) {
+                        throw new IllegalArgumentException("Blunder! " +
+                                "Declare yer event as such: 'deadline * /from * /to *', ye scurvy dog!");
+                    }
+                    LocalDateTime from = parseDate(tempArray[0].trim());
+                    LocalDateTime to = parseDate(tempArray[1].trim());
+                    printDivider();
+                    tasks.add(new Event(description, from, to));
+                    taskCount += 1;
+                    System.out.println("I've appended this to yer list:\n" + tasks.get(taskCount - 1).toString());
+                    printDivider();
+                    break;
+                }
+                default:
```
Maybe you should try to follow the `switch/case` convention in the [layout guide](https://se-education.org/guides/conventions/java/intermediate.html#layout)? I notice it in other places too.

bachletuankhai commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,402 @@
+        } catch (DateTimeParseException e) {
+            throw new IllegalArgumentException("Blunder! " +
+                    "The date format ye provided be as tangled as a ship's riggin'.\n" +
+                    "Write yer dates in the format dd/MM/yyyy HHmm or dd/MM/yyyy");
```
I like how you break a long string into multiple lines here 💯!

bachletuankhai reviewed (commented):
Overall, good job in following most of the coding convention! Try adding header comments for your classes and their public methods and fixing some minor style issues and your code will be neat!

H4mes committed:
Refactor Duke class

Split Duke class functionality into multiple subclasses.

TaskList class added to handle editing of current tasklist within the
program.

Storage class added to handle reading and writing of tasklist data to
hard disk.

Parser class added to handle interpreting of user input commands.

Ui class added to handle displaying of information.

Command abstract class and subclasses added to handle multiple possible
user commands in a streamlined manner within the main program loop.

Multiple minor edits made to Javadoc comments and formatting

H4mes committed:
Initial package restructure for better code organization

Added packages for duke, duke.task and duke.command.

Fixed TaskList(ArrayList) constructor bug where taskCount was not initialised leading to
incorrect reading and writing to file.

H4mes committed:
Resolve merge conflict

H4mes committed:
Remove unwanted test files and data directory

H4mes committed:
Merge branch 'add-gradle-support' of https://github.com/H4mes/ip

H4mes committed:
Merge add-gradle-support

H4mes committed:
Add Junit tests for Parser and Event Classes

H4mes committed:
Update build.gradle

H4mes committed:
Edit JavaDoc for TaskList Class

H4mes committed:
Add Find Functionality

Added additional logic to Parser class to handle find command inputs.
Added corresponding FindCommand class to handle execution of the find
command.
Added corresponding findTasks method to TaskList class for OOP-compliant
searching for matches.

Added new methods to Ui to allow for handling of printing lists to the
console. Modified ListCommand to utilise the new Ui method.

H4mes committed:
Clean up code to follow Java Coding Standards

H4mes committed:
Resolve merge conflicts with Level-9

H4mes committed:
Fix bug displaying 1. for all list elements

H4mes committed:
Update README.md

H4mes committed:
Update README.md

H4mes committed:
Update README.md

H4mes committed:
Update README.md

H4mes committed:
Add gradle checkstyle

Add .xml files for gradle checkstyle under config/checkstyle/ dir.

Cleaned up all syntax errors in code across multiple classes.

Removed erroneous merge conflict messages that were found in javadocs.

H4mes committed:
Implement JavaFX GUI

Changed Command and all subclasses to return a String for GUI to
display in addition to printing to the console.

Rectified bug in Storage and Tasklist instantiation and updating that
caused both private tasks arraylists to reference the same object
causing double updates to be displayed when adding tasks.

Added Launcher, Main, MainWindow and DialogBox classes to manage the
GUI.Also added DialogBox and MainWindow fxml files. Added images for
user and bot.

H4mes committed:
Enforce adherence to style guide for tests

Reorganised imports and edited code for ParserTest and EventTest classes
to pass checkstyle automated testing.

H4mes committed:
Merge branch 'master' of https://github.com/H4mes/ip into A-Assertions

H4mes committed:
Update README.md

Updated readme to reflect the completed GUI feature. Added new upcoming
feature of flexible command syntax as a checklist item.

H4mes committed:
Add assertions

Added assertions to Duke to ensure TaskList object and Storage object do
not reference the same ArrayList for each individual list of tasks.

Added assertion to parser to ensure that Event dates that start after
they end correctly throw errors.

H4mes committed:
Improve Code Quality

Refactored loadData() method in the Storage class and handleInput()
method in the Parser class to better follow SLAP guidelines. Broke both
methods up into smaller helper functions.

Rephrased several method header comments to use action words to better
follow code quality guidelines.

H4mes committed:
Remove redundant comments

Removed commented out old code from Parser class.

H4mes committed:
Merge pull request #2 from H4mes/branch-A-Assertions

Add assertions

H4mes committed:
Merge pull request #3 from H4mes/branch-A-CodeQuality

Refactor code

H4mes committed:
Add enums of alternative syntax

Added enums that allow for alternative instruction syntax and date-time
input formats.

Refactored and updated handleInput and helper functions to correctly
handle all variant inputs.

H4mes committed:
Fix incorrect DateSpecifiers

Fixed incorrect date specifiers for TO patterns.

H4mes committed:
Add help command

Added a help command for easier reference of new friendlier syntax.

Added missing documentation for code in Parser class.

H4mes committed:
Add welcome message and icon

Added icon to the GUI window. Added automated welcome message on program
start.

H4mes committed:
Clean up code to fit style guide

Corrected style errors in code.

H4mes committed:
Merge pull request #4 from H4mes/branch-C-FriendlierSyntax

Friendlier Syntax

H4mes committed:
Refactor code and polish GUI

Refactored GUI code into a package named javafx. Edited javafx code and
fxml files for a better looking GUI overall. Added styles dir with css
file for styling the scrollbar.

Refactored all mentions of Duke to correctly reference the new name of
the program as Yarr.

Added Javadoc comments for javafx classes.

Rectified a bug in "find" command that didnt match uppercase searches
with any text.

H4mes committed:
Merge pull request #5 from H4mes/branch-A-BetterGUI

Refactor code and polish GUI

H4mes committed:
Modify find command

Find command returns an ArrayList of matched tasks. Users can find it
confusing to mark or delete tasks from the provided response as their
indexes are not shown.

Modifying the find command to use an arraylist of <Task, Integer> pairs
allows for the tasks' indexes to be captured.

Let's use a new method in Ui to print these tasks with their original
index to streamlien user experience.

Using a separate method is preferable to having the displayList method
handle both standard list and find lists.

H4mes committed:
Update tests

Parser and Event tests are out of date. Parser class's parseDate in
particular does not account for friendlier syntax date formats.

Tests should account for alternative date syntax.

Adding two methods for addition alternative syntax allows us to cover
all cases where parseDate is expected to correctly produce LocalDateTime
from input.

H4mes committed:
Merge pull request #6 from H4mes/branch-A-BetterGUI

Modify find command

H4mes committed:
Merge pull request #7 from H4mes/branch-A-MoreTesting

Update tests

H4mes committed:
Rectify build.gradle

Let's have gradle correctly output yarr.jar and target the launcher as
its main class.

H4mes committed:
Merge branch 'master' of https://github.com/H4mes/ip

H4mes committed:
Rectify typo

H4mes committed:
Rectify references to duke

H4mes committed:
Update README.md

H4mes committed:
Rectify hyperlink
